pants_plugins/sendwave/pants_node/package.py

```diff
@@ -1,64 +1,141 @@
+
+    stripped_digests = []
+    for root, entries in roots_to_entries.items():
+        # create a separate digest for each source root
+        digest = await Get(Digest, CreateDigest(entries))
+        # remove the source root from the digest
+        stripped_digests.append(
+            (await Get(Digest, RemovePrefix(digest, root.path)))
+        )
+    # merge the digests together together
+    return await Get(Digest, MergeDigests(stripped_digests))
+
+
+@dataclass(frozen=True)
+class NodePathRequest:
+    """Empty request to get NodePaths.
+
+    Get requires an argument, but all our configuration is in the
+    NodeSubsystem which is injected separately. So this is just a type
+    marker to tell pants how to give us what we want.
+    """
+
+    pass
+
+
+@dataclass(frozen=True)
+class NPMPath:
+    """Path to npm executable + search path for launched processes."""
+
+    binary_path: str
+    search_paths: Tuple[str]
 
 
 @rule
-async def get_npm_package_files(request: PackageFileRequest) -> Digest:
-    project_root = PurePath(request.package_root)
-    package_json_path = project_root.joinpath("package.json")
-    package_lock = project_root.joinpath("package-lock.json")
-    yarn_lock = project_root.joinpath("yarn.lock")
-    npm_shrinkwrap = project_root.joinpath("npm-shrinkwrap.json")
-
-    rooted_configs = await Get(
-        Digest,
-        PathGlobs(
-            [
-                str(package_json_path),
-                str(package_lock),
-                str(yarn_lock),
-                str(npm_shrinkwrap),
-            ]
+async def get_node_search_paths(request: NodePathRequest,
+                                node: NodeSubsystem) -> NPMPath:
+    """Build NPMPath object from NodeSubsystem configuration."""
+    use_nvm = node.options.use_nvm
+    if use_nvm:
+        nvm_bin = await Get(Environment, EnvironmentRequest(["NVM_BIN"]))
+        if nvm_bin:
+            search_paths = (([nvm_bin["NVM_BIN"], *node.options.search_paths]))
+    else:
+        search_paths = tuple(node.options.search_paths)
+
+    npm_paths = await Get(
+        BinaryPaths,
+        BinaryPathRequest(
+            binary_name="npm",
+            search_path=search_paths,
         ),
     )
-    unrooted_configs = await Get(Digest, RemovePrefix(rooted_configs, project_root.name))
-    return unrooted_configs
+    if not npm_paths.first_path:
+        raise ValueError("Could not find npm in: ({}) cannot create package"
+                         .format(search_paths))
+    return NPMPath(binary_path=npm_paths.first_path.path,
+                   search_paths=search_paths)
 
 
 @dataclass(frozen=True)
 class NodeSourceFilesRequest:
-    package_address: str
+    """Get all tranisitvely dependent source files for given node package."""
+
+    package_address: Address
 
 
 @rule
 async def get_node_package_file_sources(
-    request: NodeSourceFilesRequest,
-) -> StrippedSourceFiles:
+        request: NodeSourceFilesRequest
```

Comment:
we used to strip the source_roots of all files, we aren't doing that anymore since instead we strip the package_root! we do that later in the main rule though